src/solvers/prop/bdd_expr.h

```diff
@@ -27,6 +27,13 @@ Author: Michael Tautschnig, michael.tautschnig@qmul.ac.uk
 
 class namespacet;
 
+/// Convert an expression to a bddt
+bddt bdd_from_expr(
+  const exprt &expr,
+  bdd_managert &bdd_mgr,
+  std::unordered_map<exprt, bddt, irep_hash> &expr_to_var,
+  std::vector<exprt> &var_to_expr);
```

Comment:
Despite the commit message I still don't understand why this is necessary or even the right thing to do. Specifically, it exposes implementation details (the map and the vector) that need to be in sync. The purpose of the `bdd_exprt` class was to encapsulate these implementation details.

Reply:
Yes, you are right this should be encapsulated. My problem with `bdd_exprt` is that it binds together the `bddt` and the `bdd_managert` meaning get only one `bddt` by manager, whereas the same manager should be shared between many BDDs. If that's fine, I will change the interface of `bdd_exprt` so that the result of the conversion is returned instead of being stored in a field.

Reply:
That would seem like a much better approach. Feel free to fiddle with the `bdd_managert` member (for example, it might become a reference member). I'm not picky about the final `bdd_exprt` interface, but I'd like to maintain appropriate encapsulation. Data structures that need to be kept in sync are bound to get us in trouble.